lib/helper.c

```diff
@@ -251,18 +251,6 @@ int fuse_parse_cmdline_30(struct fuse_args *args,
 	return rc;
 }
 
-/**
- * Compatibility ABI symbol for systems that do not support version symboling
- */
-#if (defined(__UCLIBC__) || defined(__APPLE__))
-int fuse_parse_cmdline(struct fuse_args *args,
```

Comment:
This means that when building libfuse with UCLIBC, the shared library will no longer have a `fuse_parse_cmdline` symbol. This means that programs that have been compiled against an older version of libfuse will fail when running with the newer shared object.

Am I missing something?

Reply:
Hi Nikolaus, my tired eyes might miss something myself, but this symbol is only defined for __ULIBC__ and __APPLE__ only? At least I had introduced that symbol it for these...

Anyway, given that ulibc (µClibc?) supports version symboling according to @giuliobenetti, and given that Appli is not supported anyway, all these workarounds are not needed. 

Here is the patch I wanted to create weeks ago, https://github.com/aakefbs/libfuse/commit/0cc08aba3fc4e2df644357cffcc1e5fce179ff9c 
I will review tomorrow myself, as it got rather late here. And maybe @giuliobenetti could give it a test first?

And sorry for the late patch, I had been totally occupied for the last weeks with other work (part of it for another fuse topic).

Reply:
Hmm, actually I'm not sure anymore
http://www.etalabs.net/compare_libcs.html

```
ABI and versioning comparison	musl	uClibc	dietlibc	glibc

Symbol versioning	no	no	no	yes
```

Maybe µClibc did not get updated to -ng, but there are other libc which also do not support it

Reply:
Hi @aakefbs,

sorry for absence too, I am very busy at the moment.
uclibc/-ng supports symver from a version on. That's why I would deal with HAVE_SYMVER_ATTRIBUTE instead of __UCLIBC__. The problem we're dealing with here is if the libc supports or not Symbol versioning.

If we have false positive in testing HAVE_SYMVER_ATTRIBUTE in meson then let's try to improve it.
But if I'm not wrong we also need to keep both:
fuse_session_loop_mt()
fuse_session_loop_mt_312()
because if a binary has been linked with fuse_session_loop_mt_312() with no symver, then we will end up having only
fuse_session_loop_mt() and the runtime linking could fail. Am I correct?

What do you all think about it?

This is not an easy task taking into account the ABI compatibility!

Reply:
Ah, anyway I've started the building of 44 sets of mixed toolchains with buildroot/utils/test-pkg and let's see what comes out.
There is a mix of architectures, gcc versions and libc types and versions.

Reply:
Thank you!

HAVE_SYMVER_ATTRIBUTE is just to check if the compiler supports "__attribute__((__symver__" or the old asm way should be used, see lib/fuse_misc.h. I don't think my patch to remove __UCLIBC__ and __APPLE__ would make a difference here.

Reply:
@aakefbs no because you assume we always have symver available. That's why I propose to use HAVE_SYMVER_ATTRIBUTE only. But the problem of fuse_session_loop_mt()/_312() I think remain.
@Nikratio can you give us feedback on this and this https://github.com/libfuse/libfuse/pull/698#discussion_r1015980820?

Thank you!

Reply:
Yes, we need to keep the various compatibility versions of functions around, otherwise programs compiled against old libfuse will fail when loading a new libfuse at runtime.

I think the following course of action makes most sense:

- Refuse to compile libfuse as a shared library if libc does not support symbol versioning (static linking is still possible)
- Rename functions so that the newest implementation does not have a suffix (that way it will be available when compiling a static library)
- Use symbol versioning to map older versions of a symbol to older implementations (functions with a suffix)

Would that work? Or is there some requirement that the "main" symbol (without prefix) is not defined at all? In that case I'd say we should use a preprocessor macro that is evaluated only when compiling statically.

Reply:
> Hi Nikolaus, my tired eyes might miss something myself, but this symbol is only defined for **ULIBC** and **APPLE** only? At least I had introduced that symbol it for these...

Yes. But that still means that if a program has been compiled against an old fuse version (under ulibc), and is now runtime-linked against a newer shared library (still compiled with ulibc), then it will suddenly fail.

That said, I do not know if there are any such programs. Has it ever been possible to compile libfuse dynamically against ulibc?

Reply:
> Yes, we need to keep the various compatibility versions of functions around, otherwise programs compiled against old libfuse will fail when loading a new libfuse at runtime.
> 
> I think the following course of action makes most sense:
> 
>     * Refuse to compile libfuse as a shared library if libc does not support symbol versioning (static linking is still possible)
> 
>     * Rename functions so that the newest implementation does not have a suffix (that way it will be available when compiling a static library)
> 
>     * Use symbol versioning to map older versions of a symbol to older implementations (functions with a suffix)
> 
> 
> Would that work? Or is there some requirement that the "main" symbol (without prefix) is not defined at all? In that case I'd say we should use a preprocessor macro that is evaluated only when compiling statically.

This is a solution for me.

Reply:
> > Hi Nikolaus, my tired eyes might miss something myself, but this symbol is only defined for **ULIBC** and **APPLE** only? At least I had introduced that symbol it for these...
> 
> Yes. But that still means that if a program has been compiled against an old fuse version (under ulibc), and is now runtime-linked against a newer shared library (still compiled with ulibc), then it will suddenly fail.
> 
> That said, I do not know if there are any such programs. Has it ever been possible to compile libfuse dynamically against ulibc?

Yes, uclibc dynmically  links.

Reply:
@Nikratio I don't think symbol versioning is an absolute requirement for ABI compatibility. On architechtures that do not support versioned symbols, the function is always redirected at compilation time via a macro to a versioned function name. When the function is made incompatible, a new function with a higher version in the name is created and the old function then just calls into the new one with everything to provide compatibility. I don't see how this could break something and why static compilation would be needed for libc's that do not support versioned symbols.

My assumption till yesterday night just was that the APPLE and UCLIBC macros are not required at all, as APPLE is not supported anyway and UCLIBC seemed to support versioned symbols. Now given the libc table I had found, it is really just gnu-libc that support version symbol, I *think* we should do the other way around - replace the UCLIBC and APPLE macros with a more generic way like HAVE_VERSIONED_SYMBOL and then support more libcs. Personally I actually like musl...

The only interesting part for me is, that I had created a patch last time to artifically test ulibc compilation and that had failed for me. I will check more closely now what needs to be fixed.

Reply:
@aakefbs Are you sure that the #define redirects the function name?
I mean:
```
fuse_session_loop_mt_312()
#define fuse_session_loop_mt() fuse_session_loop_mt_312()
```
should end up with only fuse_session_loop_mt_312 symbol defined and not fuse_session_loop_mt.
So some binary out there will require fuse_session_loop_mt() while other fuse_session_loop_mt_312().
#define only substitute fuse_session_loop_mt() with fuse_session_loop_mt_312() all around, including the binaries linked to
to libfuse since that #define is specified in the header.
Can you try to hint:
```
nm libfuse.x.x.so
```
and grep for both functions? If build for the case where fuse_session_loop_mt_312() is used, *without symver*, I'm not sure you will find fuse_session_loop_mt symbol, but you should find fuse_session_loop_mt_312 symbol instead.

Right?

Reply:
And instead some binary will require fuse_session_loop_mt_312() explicitly because it's defined in .c file.

Reply:
The ugly part is that other libc also define __GLIBC__ and similar macros *sigh*. Looks like we need to make it mostly a meson option, users need to define. Like here

https://lists.fedorahosted.org/archives/list/lvm2-commits@lists.fedorahosted.org/thread/OGP5IMLH7H6E4KCTB2WTLYGWPSGU6DOY/

UCLIBC can be tested for with a macro, but musl cannot, afaik. At least there needs to be a manual override.

Reply:
@giuliobenetti It is not about if the symbol is defined or not (it gets defined via the glibc version mechanism), it is about if the caller will ever use the function_name() or if the caller will use function_name_version(). With `#define function_name function_name_version` in an included header file, function_name() is always redirected. This already required for API compatibility (i.e. recompilation with FUSE_USE_VERSION XYZ, when XYZ is not the most recent version).

The entire discussion here started, as the macro redirecting fuse_parse_cmd_line was conflicting with a compat ABI symbol - fuse_parse_cmdline was not versioned so far.

Reply:
@aakefbs Yes, it was what I meant about the define, regarding the symbol you're right because of the FUSE_USE_VERSION XYZ, I only wanted to point that we need all function versions.
It looks a bit like a loop this case :-/